web3/middleware/filter.py

```diff
@@ -0,0 +1,324 @@
+from eth_utils import (
+    apply_key_map,
+    to_hex,
+    to_list,
+)
+from toolz.itertoolz import (
+    concat,
+)
+
+
+def segment_count(start, stop=None, step=5):
+    """Creates a segment counting generator
+
+    The generator returns tuple pairs of integers
+    that correspond to segments in the provided range.
+
+    :param start: The initial value of the counting range
+    :param stop: Optional, if provided the last value in the
+    counting range
+    :param step: Optional, the segment length. Default is 5.
+    :type start: int
+    :type stop: int
+    :return: returns a generator object
+
+
+    Example:
+
+    >>> segment_counter = segment_count(start=0, stop=10, step=3)
+    >>> next(segment_counter)
+    (0, 3)
+    >>> next(segment_counter)
+    (3, 6)
+    >>> next(segment_counter)
+    (6, 9)
+    >>> next(segment_counter) #  Remainder is also returned
+    (9, 10)
+    """
+    def finite(start, stop, step):
+        #  If the initial range is less than the step
+        #  just return (start, stop)
+        if stop < step:
+            yield (start, stop)
+            return
+        for i in zip(
+                range(start, stop - step, step),
+                range(start + step, stop, step)):
+            yield i
+        else:
+            remainder = (stop - start) % step
+            #  Handle the remainder
+            if remainder:
+                yield (stop - remainder, stop)
+
+    def infinite(start, step):
+        for i in zip(
+                count(start, step),
+                count(start + step, step)):
+            yield i
+
+    if stop is None:
+        return infinite(start, step)
+    else:
+        return finite(start, stop, step)
+
+
+def block_range_chunker(start_block, last_block=None, step=5):
```

Comment:
Maybe use the *segment* name here too:  `get_segmented_block_range`.

Reply:
Im using "segment" to denote this kind of output (with 1-length segments): (0, 1), (1, 2), (2, 3), (3, 4), etc.

Then there is the fromBlock -> toBlock type ranges, which the equivalent of above would be: (0, 0), (1, 1), (2, 2), etc.

Because the fromBlock/toBlock type ranges seem a little less conventional, I decided to make the basic function output the more conventional "segment" ranges (where given a range (a, b): b - a = step), and making them conform to the fromBlock/toBlock expected values as a separate step when iterating over the segments.

Chunker is confusing, because it is the same concept as segmenting.  When Im dealing with the block type ranges, I'll call them, "block ranges".